### ISSUE TITLE
Connection refused error upon starting `pscid`

### ISSUE THREAD

**Issue by dpaetzel** [kRITZCREEK/pscid#77]
First of all: pscid is awesome! Thank you for making it!

Right now, on my machine, running `pscid` fails with the following error:

```
Starting purs ide server
/nix/store/dvzm9gdda8a6vbjra63zqly74jkvdp3j-pscid-2.9.3/lib/node_modules/pscid/output/Effect.Aff/foreign.js:532
                throw util.fromLeft(step);
                ^

Error: connect ECONNREFUSED ::1:15740
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16) {
  errno: -111,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '::1',
  port: 15740
}

Node.js v18.14.2
```

Interestingly, a `purs ide server` process with the correct port configured is spawned:

```
> ps aux | grep purs
theuser    2816343  0.0  0.1 1074468872 27776 pts/8 Sl  18:30   0:00 purs ide server -p 15740 --output-directory output
```

A few months ago, running `pscid` *just worked* (I just revisited an older project of mine). My first thought was that this may have something to do with my firewall but disabling that does not change anything. Also, `nc` works:

```
> nc localhost 15740
bla
{"result":"Parsing the command failed with:\nError in $: Failed reading: not a valid json value at 'bla'\nCommand: bla","resultType":"error"}
```

To be honest, I'm a bit blank on what the problem may be. Maybe `pscid` tries to connect too fast to the `purs ide` server? Help would be much appreciated!

**Comment by kritzcreek** [comment]
Hello, thanks for the kind words!

To be honest I'm also not sure what's going on there, `pscid` or the underlying `purescript-psc-ide` library haven't changed in a long time. It's possible something made the server startup take longer and now the [100ms we wait](https://github.com/kritzcreek/purescript-psc-ide/blob/ccd4260b9b5ef8903220507719374a70ef2dd8f1/src/PscIde/Server.purs#L100) isn't enough anymore. It might be better to poll the server with commands until it responds to detect readyness...

I can offer you a workaround if that's the problem for now. You could start the server yourself and pass `-p YOUR_PORT` as an option to `pscid`. If it detects a running server with the correct working directory it'll just use that, rather than trying to start its own.

**Comment by dpaetzel** [comment]
Thank you for responding so quickly! A workaround would be great!

> start the server yourself

If I do

```
purs ide server -p 12345
```

the server process is running. If I then start a second process by doing

```
pscid -p 12345
```

I get the following output from pscid

```
Starting purs ide server
Failed to start psc-ide-server with: Error code returned: 1
```

If I understood correctly, this *should* work?

**Comment by kritzcreek** [comment]
You need to make sure to start the server in the directory of your project that holds the `output/` folder (usually the root directory of the project).

If that still doesn't work my only other guess would be that it's some IPv6 vs IPv4 difference. I'm thinking that based on the log output you pasted above. (`'::1'` is the IPv6 equivalent of `127.0.0.1`)

**Comment by dpaetzel** [comment]
> You need to make sure to start the server in the directory of your project that holds the output/ folder (usually the root directory of the project).

Yes, I did that.

> If that still doesn't work my only other guess would be that it's some IPv6 vs IPv4 difference. I'm thinking that based on the log output you pasted above. ('::1' is the IPv6 equivalent of 127.0.0.1)

OK, thank you for the hint. I can ping `::1` without problems. Interestingly, the `purs ide server` seemingly doesn't accept IPv6 connections:

Running (having running a `purs ide server -p 12345`)

```
socat -- TCP6:::1:12346
```

yields

```
2023/03/14 21:56:56 socat[1512925] E connect(5, AF=10 [0000:0000:0000:0000:0000:0000:0000:0001]:12345, 28): Connection refused
```
whereas running

```
socat - TCP4:localhost:12345
```

connects and upon entering `test` I also get a response:

```
{"result":"Parsing the command failed with:\nError in $: string at 'test'\nCommand: test","resultType":"error"}
```

My socket programming is a bit rusty so off the top of my head I don't know what exactly is going on here. I can without problems run a socat IPv6 listener and connect to that so it does not seem to be an OS problem?

**Comment by dpaetzel** [comment]
OK, I found [this thread](https://stackoverflow.com/questions/72648487/node-js-16-17-changed-resolution-of-localhost) and more specifically [this answer](https://stackoverflow.com/a/74216410). Node.js seemingly recently (sometime within the last year) changed the behaviour of `net.createConnection` (which is used by the `purescript-psc-ide` library) with respect to IPv4 vs. IPv6.

**Comment by dpaetzel** [comment]
Would it be possible to somehow synchronize the connection creation behaviour between the `purescript-psc-ide` and `pscid` code?

EDIT: I actually meant between `purescript-psc-ide` and `purs ide server`.

**Comment by dpaetzel** [comment]
`purs ide server` sets the listen address [here](https://github.com/purescript/purescript/blob/master/app/Command/Ide.hs#L46-L49) and hardcodes it to use IPv4 (they use `AF_INET` which is IPv4 whereas `AF_INET6` would be IPv6).

**Comment by dpaetzel** [comment]
You actually set the host to connect to in the `purescript-psc-ide` [code](https://github.com/kritzcreek/purescript-psc-ide/blob/master/src/PscIde.js#L8) as `127.0.0.1` and Node.js's `net.createConnection` should interpret that correctly as an IPv4 address. :thinking:

**Comment by semio** [comment]
Hey @dpaetzel @kritzcreek I encountered the same issue, and thanks to all info you provided above I think I figured out what is the problem. 

pscid is using purescript 0.13.6 package set, and the psc-ide package in the package set is too old: v15.0.1, which is from 2019 and does not set the host in [the send function ](https://github.com/kritzcreek/purescript-psc-ide/blob/0f5c433c57391ae68ce1b285a0964bd19fdae073/src/PscIde.js#L10).  Then node changed the way it handle localhost resolution, so that for some computer it will resolve to ::1. And causes the problem.

@kritzcreek Is it ok to upgrade to a newer package set? I am new to Purescript so not sure what would be affected

@dpaetzel There is a work around, you can fix the IP of localhost in your OS. For me, I edited /etc/hosts and remove localhost from the line for ::1 and this made localhost only points to 127.0.0.1 and fixed the issue.

**Comment by kritzcreek** [comment]
Hey @semio, thank you for the investigation! I don't think it should be a problem to update the package-set. I wouldn't mind making the change myself, but if you'd like to give it a try I'm happy to review your PR as well. Just let me know what you'd prefer.

**Comment by semio** [comment]
Thanks! I am glad to make a PR, will have a try at the weekend

**Comment by kritzcreek** [comment]
I'll close this after making a release that actually fixes the issue.

**Comment by kritzcreek** [comment]
Fixed in `v2.10.0`. If the issue persists or comes up again please re-open this issue. Thanks!
